src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -27372,5 +27368,554 @@ public ValueTuple(T1 item1)
+                Assert.Equal("ConstructedNamedTypeSymbol: (System.Int32 Item1, System.Int32 Item2, System.Int32 Item3, System.Int32 Item4, System.Int32 Item5, System.Int32 Item6, System.Int32 Item7, System.Int32 Item8)", print(tuple2));
+                AssertEx.SetEqual(new[] {
+                    "SubstitutedFieldSymbol: Rest",
+                    "TupleElementFieldSymbol: Item1",
+                    "TupleElementFieldSymbol: Item2",
+                    "TupleElementFieldSymbol: Item3",
+                    "TupleElementFieldSymbol: Item4",
+                    "TupleElementFieldSymbol: Item5",
+                    "TupleElementFieldSymbol: Item6",
+                    "TupleElementFieldSymbol: Item7",
+                    "TupleVirtualElementFieldSymbol: Item8"
+                }, printFields(tuple2));
+
+                var tuple3 = (NamedTypeSymbol)module.GlobalNamespace.GetMember<MethodSymbol>("C.M3").ReturnType;
+                Assert.Equal("ConstructedNamedTypeSymbol: (System.Int32 a, System.Int32 b, System.Int32 c, System.Int32 d, System.Int32 e, System.Int32 f, System.Int32 g, System.Int32 h)", print(tuple3));
+                AssertEx.SetEqual(new[] {
+                    "SubstitutedFieldSymbol: Rest",
+                    "TupleElementFieldSymbol: Item1",
+                    "TupleElementFieldSymbol: Item2",
+                    "TupleElementFieldSymbol: Item3",
+                    "TupleElementFieldSymbol: Item4",
+                    "TupleElementFieldSymbol: Item5",
+                    "TupleElementFieldSymbol: Item6",
+                    "TupleElementFieldSymbol: Item7",
+                    "TupleVirtualElementFieldSymbol: Item8",
+                    "TupleVirtualElementFieldSymbol: a",
+                    "TupleVirtualElementFieldSymbol: b",
+                    "TupleVirtualElementFieldSymbol: c",
+                    "TupleVirtualElementFieldSymbol: d",
+                    "TupleVirtualElementFieldSymbol: e",
+                    "TupleVirtualElementFieldSymbol: f",
+                    "TupleVirtualElementFieldSymbol: g",
+                    "TupleVirtualElementFieldSymbol: h", }, printFields(tuple3));
+
+                Assert.Equal("ConstructedNamedTypeSymbol: (System.Int32, System.Int32, System.Int32, System.Int32, System.Int32, System.Int32, System.Int32, System.Int32)", print(tuple3.TupleUnderlyingType));
+                AssertEx.SetEqual(new[] {
+                    "SubstitutedFieldSymbol: Rest",
+                    "TupleElementFieldSymbol: Item1",
+                    "TupleElementFieldSymbol: Item2",
+                    "TupleElementFieldSymbol: Item3",
+                    "TupleElementFieldSymbol: Item4",
+                    "TupleElementFieldSymbol: Item5",
+                    "TupleElementFieldSymbol: Item6",
+                    "TupleElementFieldSymbol: Item7",
+                    "TupleVirtualElementFieldSymbol: Item8" }, printFields(tuple3.TupleUnderlyingType));
```

Comment:
>tuple3.TupleUnderlyingType [](start = 75, length = 26)

It would be useful to assert if it is the same as one of the tuples from cases above.

Reply:
They are not reference equal, if that's what you're asking.
I don't think we expect independently constructed types to be reference equal.

---
In reply to: [426890356](https://github.com/dotnet/roslyn/pull/44231#discussion_r426890356) [](ancestors = 426890356)

Reply:
> They are not reference equal, if that's what you're asking.

No that is not what I am asking. I am asking to add asserts for relationship between the types. Can one of them be considered as an underlying tuple type for another, etc. 

---
In reply to: [426914637](https://github.com/dotnet/roslyn/pull/44231#discussion_r426914637) [](ancestors = 426914637,426890356)

Reply:
Also, are they equal in some way, etc.

---
In reply to: [426919525](https://github.com/dotnet/roslyn/pull/44231#discussion_r426919525) [](ancestors = 426919525,426914637,426890356)